hera_sim/simulate.py

```diff
@@ -0,0 +1,335 @@
+
+            # Ensure required parameters have been set.
+            assert (
+                n_freq is not None
+            ), "if data_filename not given, n_freq must be given"
+            assert (
+                n_times is not None
+            ), "if data_filename not given, n_times must be given"
+            assert (
+                antennas is not None
+            ), "if data_filename not given, antennas must be given"
+            assert (
+                ant_pairs is not None
+            ), "if data_filename not given, ant_pairs must be given"
+
+            # Default values for mutable parameters
+            if pols is None:
+                pols = ["xx"]
+            if telescope_location is None:
+                telescope_location = io.HERA_LOCATION
+            if telescope_lat_lon_alt is None:
+                telescope_lat_lon_alt = io.HERA_LAT_LON_ALT
+
+            # Actually create it
+            self.data = io.empty_uvdata(
+                nfreq=n_freq,
+                ntimes=n_times,
+                ants=antennas,
+                antpairs=ant_pairs,
+                pols=pols,
+                time_per_integ=time_per_integ,
+                min_freq=min_freq,
+                channel_bw=channel_bw,
+                instrument=instrument,
+                telescope_location=telescope_location,
+                telescope_lat_lon_alt=telescope_lat_lon_alt,
+                object_name=object_name,
+                start_jd=start_jd,
+                vis_units=vis_units,
+            )
+
+        else:
+            self.data = self._read_data(self.data_filename, pols, **kwargs)
+
+    def _read_data(self, filename, polarizations, **kwargs):
+        uv = UVData()
+        uv.read(filename, polarizations=polarizations, read_data=True, **kwargs)
+        return uv
+
+    def write_data(self, filename, file_type=None, **kwargs):
+        """
+        Write current UVData object to file.
+
+        Args:
+            filename (str): filename to write to. Suffix of filename will determine the type of file to be written,
+                unless `file_type` is set.
+            file_type: (str): one of "miriad", "uvfits" or "uvh5" (i.e. any of the supported write methods of
+                :class:`pyuvdata.UVData`) which determines which write method to call.
+            **kwargs: keyword arguments sent directly to the write method chosen.
+        """
+        if file_type is None:
+            try:
+                file_type = filename.split(".")[-1]
+            except IndexError:
+                raise ValueError(
+                    "if no file_type is given, the filename must be suffixed with either .miriad, .uvfits or .uvh5"
+                )
+
+        if file_type not in ["miriad", "uvfits", "uvh5"]:
+            raise ValueError("file_type must be one of 'miriad', 'uvfits' or 'uvh5'")
+
+        getattr(self.data, "write_%s" % filename.split(".")[-1])(filename, **kwargs)
```

Comment:
looks like this should actually be
```
getattr(self.data, "write_%s" % filetype)(filename, **kwargs)
```
This should have failed but you got lucky b/c your test was using a `uvh5` file whose suffix happens to also be the file type. This isn't true for `miriad`.

Reply:
Thanks for picking this up. I've changed this in 47c08ac73ed6c8ffe9243b5727ae57209bdb01dd. Now, if the user passes a file with suffix ".uv" it will automatically interpret as miriad, whereas .uvfits and .uvh5 will obviously by uvfits or uvh5. And it will actually use the assigned file_type, rather than the suffix itself.

Reply:
Hmm, an interesting idea, but I would be a little hesitant to automatically infer the filetype from the suffix, b/c it is in principle completely up to the user to define this--and someone could call a `uvfits` file something like`uvf`. Furthermore, a `uvh5` file is just an `HDF5` file which could be called `.h5` or `.hdf5`, and additionally we often have miriad files that *dont* have the suffix `uv`, but something more complicated, like `uvOCSL` or `uvRX` etc.

My opinion would be to keep the `filetype` kwarg to determine which `read_` func to call, which should default to `uvh5`. Perhaps @jaguirre has thoughts on this too...

Reply:
I don't mean to butt in too much, but I saw this comment and thought I'd say that we've done something like this in the generic read function in pyuvdata: https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/29d4a0802ea24a46b551cc55624bd332fbcec563/pyuvdata/uvdata.py#L2331

Basically, we test if the "file" is actually a directory (which means it's a miriad dataset or measurement set), and look for a `vartable` (miriad) or `OBSERVATION` (measurement set) file in the directory. For uvfits and uvh5, we look at the file suffix, but you could imagine doing something more sophisticated like doing a try-except block where you open and access the file using h5py or astropy.io.fits. I know h5py checks the file's "magic number", and can immediately tell if it's an HDF5 file or not.

Mostly just wanted to say that there's precedent for inferring file type from the name, but it's obviously not foolproof :)

Reply:
OK, I've done the easy thing and now it just ignores the file extension altogether, and uses file_type instead (which defaults to uvh5). I think this is just easier and less error prone.